Sources/UrlFormEncoding/UrlFormDecoder.swift

```diff
@@ -0,0 +1,629 @@
+      defer { self.decoder.containers.removeLast() }
+      return UnkeyedContainer(decoder: self.decoder, container: container, codingPath: self.codingPath)
+    }
+
+    mutating func superDecoder() throws -> Decoder {
+      guard !self.isAtEnd else { throw Error.decodingError("Unkeyed container is at end", self.codingPath) }
+      defer { self.codingPath.removeLast() }
+      let container = self.container[self.currentIndex]
+      self.currentIndex += 1
+      let decoder = UrlFormDecoder()
+      decoder.containers = [container]
+      decoder.codingPath = self.codingPath
+      return decoder
+    }
+  }
+
+  struct SingleValueContainer: SingleValueDecodingContainer {
+    let decoder: UrlFormDecoder
+    let container: Any
+
+    let codingPath: [CodingKey] = []
+
+    private func unwrap<T>(_ block: (String) -> T?, _ line: UInt = #line) throws -> T {
+      guard
+        let container = self.container as? String,
+        let value = block(container)
+        else { throw Error.decodingError("Expected \(T.self), got nil", self.codingPath) }
+
+      return value
+    }
+
+    func decodeNil() -> Bool {
+      return (self.container as? String)?.isEmpty ?? false
+    }
+
+    func decode(_ type: Bool.Type) throws -> Bool {
+      return try self.unwrap(Bool.init)
+    }
+
+    func decode(_ type: Int.Type) throws -> Int {
+      return try self.unwrap(Int.init)
+    }
+
+    func decode(_ type: Int8.Type) throws -> Int8 {
+      return try self.unwrap(Int8.init)
+    }
+
+    func decode(_ type: Int16.Type) throws -> Int16 {
+      return try self.unwrap(Int16.init)
+    }
+
+    func decode(_ type: Int32.Type) throws -> Int32 {
+      return try self.unwrap(Int32.init)
+    }
+
+    func decode(_ type: Int64.Type) throws -> Int64 {
+      return try self.unwrap(Int64.init)
+    }
+
+    func decode(_ type: UInt.Type) throws -> UInt {
+      return try self.unwrap(UInt.init)
+    }
+
+    func decode(_ type: UInt8.Type) throws -> UInt8 {
+      return try self.unwrap(UInt8.init)
+    }
+
+    func decode(_ type: UInt16.Type) throws -> UInt16 {
+      return try self.unwrap(UInt16.init)
+    }
+
+    func decode(_ type: UInt32.Type) throws -> UInt32 {
+      return try self.unwrap(UInt32.init)
+    }
+
+    func decode(_ type: UInt64.Type) throws -> UInt64 {
+      return try self.unwrap(UInt64.init)
+    }
+
+    func decode(_ type: Float.Type) throws -> Float {
+      return try self.unwrap(Float.init)
+    }
+
+    func decode(_ type: Double.Type) throws -> Double {
+      return try self.unwrap(Double.init)
+    }
+
+    func decode(_ type: String.Type) throws -> String {
+      return try self.unwrap(id)
+    }
+
+    func decode<T>(_ type: T.Type) throws -> T where T: Decodable {
+      self.decoder.containers.append(self.container)
+      defer { self.decoder.containers.removeLast() }
+      return try self.decoder.unbox(self.container, as: T.self)
+    }
+  }
+
+  public enum DataDecodingStrategy {
```

Comment:
These decoding strategies likewise mirror `JSONDecoder`'s, with one difference: `JSONDecoder.custom` is `(Decoder) throws -> Data`, while we're `(String) -> Data?`. We lose some error messaging ability (which we could regain with `(String) throws -> Data`, but things are generally more flexible. We can always capture a custom decoder locally and pass it our string.